torch/csrc/autograd/engine.cpp

```diff
@@ -179,17 +177,33 @@ auto ReadyQueue::pushShutdownTask() -> void {
   not_empty_.notify_one();
 }
 
-auto ReadyQueue::pop() -> NodeTask {
+auto ReadyQueue::pop() -> std::unique_ptr<NodeTask> {
   // Lock mutex for accesses to heap_
   std::unique_lock<std::mutex> lock(mutex_);
   not_empty_.wait(lock, [this]{ return !heap_.empty(); });
   // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
-  auto task = std::move(const_cast<NodeTask&>(heap_.top())); heap_.pop();
+  auto task = make_unique<NodeTask>(std::move(const_cast<NodeTask&>(heap_.top())));
```

Comment:
Basic cpp question: Is it fine to take a stack allocated object and stuff it in a unique pointer like that (wrt lifetime of the object)?

Reply:
So what this line does is:
- The heap_ contains a moved stack object.
- top() returns a const reference
- const_cast removes the const
- move this stack object to the make_unique
- make_unique creates a heap copy of that moved reference?

Is that accurate?

Reply:
> So what this line does is:

> * make_unique creates a heap copy of that moved reference?
> 
> Is that accurate?

My understanding of [make_unique](https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique) here is that we move the stack object to heap by invoking `NodeTask`s default move constructor. So this object will be in heap instead of stack after the `make_unique`. In the cpp doc it says `make_unique` equivalent to `unique_ptr<T>(new T(std::forward<Args>(args)...))`

Reply:
@smessmer could you help on giving some insights here? We couldn't be sure about whether this is a valid C++ pattern.

Reply:
Yes, your analysis sounds correct. It calls the move constructor and creates a new NodeTask instance on the heap. Note, however, that the `std::priority_queue` might not expect things being moved out from it. You're violating a class invariant of `std::priority_queue`. It would probably work with most compilers (at least as long as you make sure that the moved-from `NodeTask` object that remains in the queue is still valid to be passed to `CompareNodeTaskTime`), but there might be compilers, systems or standard library implementations where your code breaks.

Reply:
So this violation of `class invariant of std::priority_queue` seems already exist even before this PR, and the unique_ptr ownership (lifetime) should be fine if we create a new NodeTask instance on the heap with the default move contructor.